clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1599,6 +1599,8 @@ public void handleResponse(AbstractResponse response) {
                 fatalError(Errors.PRODUCER_FENCED.exception());
             } else if (error == Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED) {
                 fatalError(error.exception());
+            } else if (error == Errors.INVALID_TXN_STATE) {
```

Comment:
I went down a very long rabbit hole to find that we only returned INVALID_TXN_STATE for a short time from this request: https://github.com/apache/kafka/commit/1f2451d4e7e3766540d3650d177e304fcddf49b8 (here's the commit that removed this error). We never removed it from the errors returned I suppose because some really old broker could return it. 

Having this here doesn't hurt, I was just curious about the history behind it. 😅

Reply:
Thanks for digging up the details!